sdks/python/apache_beam/examples/complete/autocomplete.py

```diff
@@ -40,7 +40,7 @@ def run(argv=None):
   # We use the save_main_session option because one or more DoFn's in this
   # workflow rely on global context (e.g., a module imported at module level).
   pipeline_options = PipelineOptions(pipeline_args)
-  pipeline_options.view_as(SetupOptions).save_main_session = True
+  pipeline_options.view_as(SetupOptions).save_main_session = save_main_session
```

Comment:
Could you please help me understand why save_main_session is an issue here? 
Which runner has an issue if we don't set this to `False`?
Adding this flag adds some confusion to the example, as seeing this pattern users might try to copy it into their pipelines. I'd like to understand if we can avoid defining the flag as an additional parameter to the example. For example, the comment above is going out of sync with this line now.

Reply:
This is the error I got when save_main_session is set to True https://ci-beam.apache.org/job/beam_PostCommit_Python_Examples_Direct/13/testReport/junit/apache_beam.examples.complete.autocomplete_test/AutocompleteTest/test_autocomplete_output_files_on_small_input/

Reply:
for posterity as Jenkins logs don't last:
```
Error Message
RecursionError: maximum recursion depth exceeded while calling a Python object
Stacktrace
self = <apache_beam.examples.complete.autocomplete_test.AutocompleteTest testMethod=test_autocomplete_output_files_on_small_input>

    @pytest.mark.examples_postcommit
    def test_autocomplete_output_files_on_small_input(self):
      EXPECTED_PREFIXES = "t: [(3, 'to'), (2, 'this'), (1, 'that')]\n" \
                            "th: [(2, 'this'), (1, 'that')]\n" \
                            "thi: [(2, 'this')]\n" \
                            "this: [(2, 'this')]\n" \
                            "tha: [(1, 'that')]\n" \
                            "that: [(1, 'that')]\n" \
                            "to: [(3, 'to')]"
    
      # Setup the files with expected content.
      temp_folder = tempfile.mkdtemp()
      self.create_file(
          os.path.join(temp_folder, 'input.txt'), ' '.join(self.WORDS))
      autocomplete.run([
          '--input=%s/input.txt' % temp_folder,
          '--output',
>         os.path.join(temp_folder, 'result')
      ])

apache_beam/examples/complete/autocomplete_test.py:99: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
apache_beam/examples/complete/autocomplete.py:56: in run
    | 'write' >> WriteToText(known_args.output))
apache_beam/pipeline.py:596: in __exit__
    self.result = self.run()
apache_beam/pipeline.py:570: in run
    pickler.dump_session(os.path.join(tmpdir, 'main_session.pickle'))
apache_beam/internal/pickler.py:61: in dump_session
    return desired_pickle_lib.dump_session(file_path)
apache_beam/internal/dill_pickler.py:307: in dump_session
    dill.load_session(file_path)
../../build/gradleenv/1398941891/lib/python3.7/site-packages/dill/_dill.py:368: in load_session
    module = unpickler.load()
../../build/gradleenv/1398941891/lib/python3.7/site-packages/dill/_dill.py:472: in load
    obj = StockUnpickler.load(self)
../../build/gradleenv/1398941891/lib/python3.7/site-packages/execnet/gateway_base.py:143: in __getattr__
    locs = self._importdef.get(name)
../../build/gradleenv/1398941891/lib/python3.7/site-packages/execnet/gateway_base.py:143: in __getattr__
    locs = self._importdef.get(name)
../../build/gradleenv/1398941891/lib/python3.7/site-packages/execnet/gateway_base.py:143: in __getattr__
    locs = self._importdef.get(name)
E   RecursionError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)
```

Reply:
cc: @ryanthompson591 FYI. With CloudPickle we should be able to  remove such save_main_session cruft.

Reply:
@benWize is the TestDirectRunner the only one affected by it? If so, we could potentially override this setting in https://github.com/apache/beam/blob/5e528cf4889b2393703ba00e2db8fad0e406b730/sdks/python/apache_beam/runners/direct/test_direct_runner.py#L42. Otherwise I guess we can go as is, unless @ryanthompson has some other ideas to avoid the awkwardness of this flag until CloudPickle helps us get rid of it. perhaps we can still comment that branching is needed for testing only, but in the examples themselves users would pass True

Reply:
Yes, Once we make cloudpickle the default pickler all save_main_session flags can be removed. I hope to do that this quarter.

Reply:
I added a comment to specify save_main_session=False is used only for testing, also I checked that this fails only for TestDirectRunner, tested for [flink](https://ci-beam.apache.org/job/beam_PostCommit_Python_Examples_Flink/7/) and [spark](https://ci-beam.apache.org/job/beam_PostCommit_Python_Examples_Spark/9/).

Reply:
I think when I tried the example on a regular direct runner, there were also no issues. If that's the case, could you please check if we could instead remove the `--save_main_session=True` from `options` within test_direct_runner instead of having to add `--save_main_session=False` into the tests?

Reply:
I tried this but, the tests were still failing, I realized that the error is only present when running with pytest + xdist (parallel running), not sure why, but I changed the configuration of the gradle example task for direct runner and all tests are passing now.